sdk/eventhub/Microsoft.Azure.WebJobs.Extensions.EventHubs/src/Listeners/EventHubListener.PartitionProcessor.cs

```diff
@@ -0,0 +1,344 @@
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Messaging.EventHubs;
+using Azure.Messaging.EventHubs.Processor;
+using Microsoft.Azure.WebJobs.EventHubs.Processor;
+using Microsoft.Azure.WebJobs.Host.Executors;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json;
+using Azure.Core.Diagnostics;
+
+namespace Microsoft.Azure.WebJobs.EventHubs.Listeners
+{
+    internal sealed partial class EventHubListener
+	{
+		// We get a new instance each time Start() is called.
+		// We'll get a PartitionProcessor instance per partition - they will run in parallel on a single machine.
+		internal class PartitionProcessor : IEventProcessor, IDisposable
+		{
+			private readonly ITriggeredFunctionExecutor _executor;
+			private readonly bool _singleDispatch;
+			private readonly ILogger _logger;
+			private readonly CancellationTokenSource _cts = new CancellationTokenSource();
+			private readonly int _batchCheckpointFrequency;
+			private int _batchCounter;
+			private bool _minimumBatchesEnabled;
+			private bool _disposed;
+			private Task _cachedEventsBackgroundTask;
+			private TimeSpan _maxWaitTime;
+			private ValueStopwatch _currentCycle;
+			private EventProcessorHostPartition _mostRecentPartitionContext;
+			private CancellationTokenSource _cachedEventsBackgroundTaskCts;
+
+			internal PartitionProcessorEventsManager CachedEventsManager { get; }
+
+			public PartitionProcessor(EventHubOptions options, ITriggeredFunctionExecutor executor, ILogger logger, bool singleDispatch)
+			{
+				_executor = executor;
+				_singleDispatch = singleDispatch;
+				_batchCheckpointFrequency = options.BatchCheckpointFrequency;
+				_logger = logger;
+				_minimumBatchesEnabled = options.MinEventBatchSize > 1; // 1 is the default
+				CachedEventsManager = new PartitionProcessorEventsManager(maxBatchSize: options.MaxEventBatchSize, minBatchSize: options.MinEventBatchSize);
+				_maxWaitTime = options.MaxWaitTime;
+			}
+
+			public Task CloseAsync(EventProcessorHostPartition context, ProcessingStoppedReason reason)
+			{
+				// signal cancellation for any in progress executions
+				_cts.Cancel();
+				CachedEventsManager.PartitionClosing();
+
+				_logger.LogDebug(GetOperationDetails(context, $"CloseAsync, {reason.ToString()}"));
+				return Task.CompletedTask;
+			}
+
+			public Task OpenAsync(EventProcessorHostPartition context)
+			{
+				_logger.LogDebug(GetOperationDetails(context, "OpenAsync"));
+				return Task.CompletedTask;
+			}
+
+			public Task ProcessErrorAsync(EventProcessorHostPartition context, Exception error)
+			{
+				string errorDetails = $"Processing error (Partition Id: '{context.PartitionId}', Owner: '{context.Owner}', EventHubPath: '{context.EventHubPath}').";
+
+				Utility.LogException(error, errorDetails, _logger);
+
+				return Task.CompletedTask;
+			}
+
+			public async Task ProcessEventsAsync(EventProcessorHostPartition context, IEnumerable<EventData> messages, CancellationToken processingCancellationToken)
```

Comment:
I believe we should consider extracting the `lock` out of the events manager and using it here instead.   This would allow the lock to be taken at the beginning of the `ProcessEventsAsync` call and avoid having to take/release it multiple times.   Likewise, the `MonitorCachedEvents` method could take the lock before L209, which would prevent the handler from running as the background task emits the batch.

I think this provides a cleaner and more intuitive set of guarantees, as the partition processor has a better understanding of where concurrency can occur and how coarse-grained the locking should be.